src/vaultRegistry/vault/getVault.ts

```diff
@@ -0,0 +1,313 @@
+    },
+    {
+      ...vaultContract,
+      functionName: 'balanceOf',
+      args: [account]
+    }
+  ]
+}
+
+function prepareRegistryContract(address: Address, vault: Address): ReadContractParameters {
+  const vaultRegisty = {
+    address,
+    abi: VaultRegistyAbi
+  }
+
+  return {
+    ...vaultRegisty,
+    functionName: 'metadata',
+    args: [vault]
+  }
+}
+
+function prepareTokenContracts(address: Address, account: Address): ReadContractParameters[] {
+  const token = {
+    address,
+    abi: ERC20Abi
+  }
+  return [
+    {
+      ...token,
+      functionName: 'name',
+    },
+    {
+      ...token,
+      functionName: 'symbol',
+    },
+    {
+      ...token,
+      functionName: 'balanceOf',
+      args: [account]
+    }
+  ]
+}
+
+export async function getVaults({ vaults, account = ADDRESS_ZERO, owner, client }: { vaults: Address[], account?: Address, owner?: Address, client: PublicClient }): Promise<VaultData[]> {
+  const hasAccount = account !== ADDRESS_ZERO
+  const chainId = client.chain?.id as number
+
+  // Add vaultName and metadata
+  const registryMetadata = await client.multicall({
+    // @ts-ignore
+    contracts: vaults.map(vault => prepareRegistryContract(VaultRegistryByChain[chainId], vault)).flat(),
+    allowFailure: false
+  }) as unknown as string[][]
+
+  let result: any[] = registryMetadata.map((entry, i) => {
+    return {
+      address: getAddress(vaults[i]),
+      metadata: {
+        creator: getAddress(entry[2]),
+        cid: entry[3] as string,
+      }
+    }
+  })
+  if (owner) result = result.filter(entry => entry.metadata.creator === getAddress(owner))
+
+  // Get core metadata
+  const vaultData = await client.multicall({
+    // @ts-ignore
+    contracts: vaults.map(vault => prepareVaultContract(vault, account)).flat(),
+    allowFailure: false
+  })
+
+  // Add core metadata
+  result = vaultData.map((vault, i) => {
+    if (i > 0) i = i * 10
+    const assetsPerShare = Number(vaultData[i + 6]) > 0 ? Number(vaultData[i + 5]) / Number(vaultData[i + 6]) : Number(1)
+    const fees = vaultData[i + 7] as [BigInt, BigInt, BigInt, BigInt]
+    return {
+      ...result[i],
+      vault: {
+        address: getAddress(result[i].address),
+        name: String(vaultData[i + 0]),
+        symbol: String(vaultData[i + 1]),
+        decimals: Number(vaultData[i + 2]),
+        logoURI: "/images/tokens/pop.svg",
+        balance: hasAccount ? Number(vaultData[i + 9]) : 0
+      },
+      assetAddress: getAddress(vaultData[i + 3] as string),
+      adapterAddress: getAddress(vaultData[i + 4] as string),
+      totalAssets: Number(vaultData[i + 5]),
+      totalSupply: Number(vaultData[i + 6]),
+      assetsPerShare: assetsPerShare,
+      fees: {
+        deposit: Number(fees[0]),
+        withdrawal: Number(fees[1]),
+        management: Number(fees[2]),
+        performance: Number(fees[3]),
+      },
+      depositLimit: Number(vaultData[i + 8]),
+      chainId: chainId
+    }
```

Comment:
this is really ugly and prone to errors in the future. If we add another view call to the `prepareVaultContract()` function, the whole value matching here will break.

We could rewrite the function so that each vault is handled separately. Then the array of data we get from the multicall only belongs to one of the vaults. That will increase the number of rpc calls tho. I'd be fine with it if it improves readability.